vcf/toVariant.go

```diff
@@ -0,0 +1,136 @@
+// variant struct. Even if the original Vcf record is not used, the Records
+// channel must be read else the sending goroutine will be blocked.
+type SmallVariantChans struct {
+	Substitutions chan variant.Substitution
+	Insertions    chan variant.Insertion
+	Deletions     chan variant.Deletion
+	Delins        chan variant.Delins
+	Records       chan Vcf
+}
+
+// NewSmallVariantChans creates a series of channels to send the variant
+// representation of a given vcf record. The input buffSize determines
+// the channel buffer size for all on the variant channels.
+func NewSmallVariantChans(buffSize int) SmallVariantChans {
+	var answer SmallVariantChans
+	answer.Substitutions = make(chan variant.Substitution, buffSize)
+	answer.Insertions = make(chan variant.Insertion, buffSize)
+	answer.Deletions = make(chan variant.Deletion, buffSize)
+	answer.Delins = make(chan variant.Delins, buffSize)
+	answer.Records = make(chan Vcf, buffSize)
+	return answer
+}
+
+// ChanToVariants splits an incoming channel of VCFs to a channel of variant types.
+// Must be run as a goroutine else the thread will deadlock.
+// See documentation for SmallVariantChans for more information on variant channels.
+func ChanToVariants(c <-chan Vcf, sendChans SmallVariantChans) {
+	for v := range c {
+		if !canParseToBases(v.Ref) {
+			continue // cannot parse vcf record so skip
+		}
+		var refSeq, altSeq []dna.Base
+		refSeq = dna.StringToBases(v.Ref)
+		for altIdx := range v.Alt {
+			if !canParseToBases(v.Alt[altIdx]) {
+				continue // cannot parse this allele so skip
+			}
+			altSeq = dna.StringToBases(v.Alt[altIdx])
+			sendVariant(sendChans, v.Chr, v.Pos, refSeq, altSeq)
+			sendChans.Records <- v
+		}
+	}
+	closeSmallVariantChans(sendChans)
+}
+
+// GoChanToVariants wraps ChanToVariants and handles
+// variant channel creation and goroutine spawning.
+func GoChanToVariants(c <-chan Vcf) SmallVariantChans {
+	answer := NewSmallVariantChans(cap(c))
+	go ChanToVariants(c, answer)
+	return answer
+}
+
+// sendVariant determines if a single alt allele derived from a vcf record is a
+// Substitution, Insertion, Deletion, or Delins; parses the data to the appropriate
+// variant type, and sends on the respective channel.
+func sendVariant(sendChans SmallVariantChans, chr string, pos int, refSeq []dna.Base, altSeq []dna.Base) {
+	var matchingOffset int
+	refSeq, altSeq, matchingOffset = trimMatchingBases(refSeq, altSeq)
+	pos += matchingOffset
+	pos -= 1 // from 1-base to 0-base
+
+	switch {
+	case len(refSeq) == 1 && len(altSeq) == 1: // Substitution
+		sendChans.Substitutions <- variant.Substitution{Chr: chr, Pos: pos, Ref: refSeq[0], Alt: altSeq[0]}
+
+	case len(refSeq) < len(altSeq) && len(refSeq) == 0: // Insertion
+		// e.g. ATG -> ATCG
+		// VCF: pos = 2; ref = T; alt = TC
+		// Desired: pos = 2; insSeq = C
+		sendChans.Insertions <- variant.Insertion{Chr: chr, Pos: pos, Seq: altSeq}
+
+	case len(refSeq) > len(altSeq) && len(altSeq) == 0: // Deletion
```

Comment:
Some comment as for insertion.  I am guessing there is something trimming the vcf strings before sending them to this func, but maybe the comments should change to reflect this?

Reply:
These bases get trimmed inside this function. This is important as multi-allelic sites can give some ref/alt sets where there are multiple matching bases which significantly complicate the logic if not removed
e.g.
ref = ACG
alt = A, ACGCG